src/de.rs

```diff
@@ -348,19 +365,32 @@ impl<'de> de::Deserializer<'de> for Deserializer {
         }
     }
 
-    // Define real `i64` / `u64` deserializers that try to cast from `f64`.
-
+    // Define real `i64` / `u64` deserializers that only work with bigints.
     fn deserialize_i64<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
-        match self.as_safe_integer() {
-            Some(v) => visitor.visit_i64(v),
-            None => self.invalid_type(visitor),
+        if !self.value.is_bigint() {
+            return self.invalid_type(visitor);
+        }
+        let big_int: BigInt = self.value.clone().into();
+        // Safe to do unwraps since we know the type is a bigint
+        let rust_string = big_int.to_string(10).unwrap().as_string().unwrap();
```

Comment:
Hmm it's pretty wasteful to go via string conversion/passing/parsing just to pass numbers. Since there's no built-in API in BigInt to get 64-bit integer yet, I'd declare my own binding here to get number directly and use that.

Reply:
I do not think there is any way to do this safely. When converting an i64 into a JsValue it goes first into a string https://rustwasm.github.io/wasm-bindgen/api/src/wasm_bindgen/lib.rs.html#853. I think the same has to be done in the reverse. 

A possible optimization is to first check if the number is a safe integer and use the normal f64 conversion if so.

Reply:
Woah that's bad.

Emscripten does this by either passing 2x 32-bit integers (in standard mode) or just by passing 64-bit integer directly (when using `-s WASM_BIGINT`, aka https://github.com/WebAssembly/JS-BigInt-integration proposal). I think this should be doable in wasm-bindgen as well, either by making a PR upstream or doing something similar here inline.

Do you want to give it a try to make an upstream PR? If not, I might give it a go.

Reply:
> Do you want to give it a try to make an upstream PR? If not, I might give it a go.

You know what, I'm now tempted enough that I'll give it a go.

Reply:
Or maybe for now it's better to do it here and then upstream. Anyway, my idea is that you can add bindings like this:

```rust
#[wasm_bindgen]
extern {
  #[wasm_bindgen(js_name = BigInt)]
  fn from_u64(x: u64) -> BigInt;

  #[wasm_bindgen(js_name = BigInt)]
  fn from_i64(x: i64) -> BigInt;

  ...
}
```

Those will bind to the JS `BigInt` function, but use wasm-bindgen's native support for passing 64-bit integers, avoiding string representation altogether, so `BigInt` function will behave just like a no-op identity function, retrieving bigint and returning one.

Reply:
Ok, I made those changes. The bindings seem to work and do not lose any precision. Do you know if wam_bindgen's function support for u64/i64 uses strings or the native BigInt support?

> Do you want to give it a try to make an upstream PR? If not, I might give it a go.

I think you probably should do that as I am much less familiar with wasm_bindgen and the supporting infrastructure (js_sys).

Reply:
It doesn't use strings, but doesn't use native BigInt support either yet AFAIK - instead, it passes two 32-bit integers and reassembles them on the other side. This should be still a lot faster than string conversion & allocation.